test/e2e/cache/cache_server_test.go

```diff
@@ -0,0 +1,504 @@
+	return unstructured, nil
+}
+
+// TODO: refactor with TestAllScenariosAgainstStandaloneCacheServer
+func TestCacheServerAllScenarios(t *testing.T) {
+	_, dataDir, err := framework.ScratchDirs(t)
+	require.NoError(t, err)
+
+	cacheServerPortStr, err := framework.GetFreePort(t)
+	require.NoError(t, err)
+	cacheServerPort, err := strconv.Atoi(cacheServerPortStr)
+	require.NoError(t, err)
+	cacheServerOptions := cacheopitons.NewOptions(path.Join(dataDir, "cache"))
+	cacheServerOptions.SecureServing.BindPort = cacheServerPort
+	cacheServerEmbeddedEtcdClientPort, err := framework.GetFreePort(t)
+	require.NoError(t, err)
+	cacheServerEmbeddedEtcdPeerPort, err := framework.GetFreePort(t)
+	require.NoError(t, err)
+	cacheServerOptions.EmbeddedEtcd.ClientPort = cacheServerEmbeddedEtcdClientPort
+	cacheServerOptions.EmbeddedEtcd.PeerPort = cacheServerEmbeddedEtcdPeerPort
+	cacheServerCompletedOptions, err := cacheServerOptions.Complete()
+	require.NoError(t, err)
+	if errs := cacheServerCompletedOptions.Validate(); len(errs) > 0 {
+		require.NoError(t, apimachineryerrors.NewAggregate(errs))
+	}
+	cacheServerConfig, err := cacheserver.NewConfig(cacheServerCompletedOptions, nil)
+	require.NoError(t, err)
+	cacheServerCompletedConfig, err := cacheServerConfig.Complete()
+	require.NoError(t, err)
+	ctx, cancelFunc := context.WithCancel(context.Background())
+	t.Cleanup(cancelFunc)
+	if cacheServerCompletedConfig.EmbeddedEtcd.Config != nil {
+		t.Logf("Starting embedded etcd for the cache server")
+		require.NoError(t, embeddedetcd.NewServer(cacheServerCompletedConfig.EmbeddedEtcd).Run(ctx))
+	}
+	cacheServer, err := cacheserver.NewServer(cacheServerCompletedConfig)
+	require.NoError(t, err)
+	preparedCachedServer, err := cacheServer.PrepareRun(ctx)
+	require.NoError(t, err)
+	t.Logf("Starting the cache server")
+	go func() {
+		require.NoError(t, preparedCachedServer.Run(ctx))
+	}()
+	cacheServerKubeConfig := clientcmdapi.Config{
+		Clusters: map[string]*clientcmdapi.Cluster{
+			"cache": {
+				Server:               fmt.Sprintf("https://localhost:%s", cacheServerPortStr),
+				CertificateAuthority: path.Join(dataDir, "cache", "apiserver.crt"),
+			},
+		},
+		Contexts: map[string]*clientcmdapi.Context{
+			"cache": {
+				Cluster: "cache",
+			},
+		},
+		CurrentContext: "cache",
+	}
+
+	cacheClientConfig := clientcmd.NewNonInteractiveClientConfig(cacheServerKubeConfig, "cache", nil, nil)
+	cacheClientRestConfig, err := cacheClientConfig.ClientConfig()
+	require.NoError(t, err)
+	t.Logf("waiting for the cache server at %v to become ready", cacheClientRestConfig.Host)
+	waitUntilCacheServerIsReady(ctx, t, cacheClientRestConfig)
+	cacheClientRT := cacheClientRoundTrippersFor(cacheClientRestConfig)
+
+	for _, scenario := range scenarios {
+		t.Run(scenario.name, func(tt *testing.T) {
+			scenario.work(ctx, tt, cacheClientRT, logicalcluster.New("acme"), schema.GroupVersionResource{Group: "apis.kcp.dev", Version: "v1alpha1", Resource: "apiexports"})
```

Comment:
You need a `tt.Parallel()` here

Reply:
Also, comment - if you use the variable `t` for the inner `*testing.T`, it is not possible to accidentally call the upper `*testing.T`, since it is shadowed. In the past these tests have called the wrong `*testing.T`. Let's please just use `t` since it cannot be incorrect.

Reply:
yeah, that is true, I like to use `tt` just to highlight it is not `t`

also, I'm not convinced we need `tt.Parallel()` here, these tests are quick (a lot of them are under `100`ms), they just send a request to a local server.